public/app/core/utils/EventsProvider.tsx

```diff
@@ -0,0 +1,101 @@
+import React, { ComponentType, FunctionComponent, useEffect } from 'react';
+import { Observable, Subject, Unsubscribable } from 'rxjs';
+import { filter } from 'rxjs/operators';
+import { AppEvent, Subtract } from '@grafana/data';
+
+import appEvents from '../app_events';
+
+export interface EventsContextApi<SubscribeEvents, PublishEvents> {
+  events: (originFilter: string) => Observable<AppEvent<SubscribeEvents>>;
+  publish: (event: AppEvent<PublishEvents>, origin: string, payload?: PublishEvents) => void;
+}
+
+export interface RootEventsContextApi {
+  rootEvents: Observable<AppEvent<any>>;
+  rootPublish: (event: AppEvent<any>, payload?: any) => void;
+}
+
+const events = new Subject<AppEvent<any>>();
+const publish = <T extends {} = {}>(event: AppEvent<T>, payload?: T) => {
+  events.next({ ...event, payload });
+  appEvents.emit(event, payload);
+};
+
+export const rootEventing: RootEventsContextApi = {
+  rootEvents: events,
+  rootPublish: publish,
+};
+
+export const EventsContext = React.createContext(rootEventing);
+EventsContext.displayName = 'EventsContext';
+
+export const EventsProvider = ({ children }: { children: React.ReactNode }) => {
+  return <EventsContext.Provider value={rootEventing}>{children}</EventsContext.Provider>;
+};
+
+export const provideEvents = (component: ComponentType<any>) => (props: any) => (
+  <EventsProvider>{React.createElement(component, { ...props })}</EventsProvider>
+);
+
+export const withEvents = <SubscribeEvents extends {}, PublishEvents extends {}>(
+  ...subscribeEvents: Array<AppEvent<SubscribeEvents>>
+) => {
+  type ComponentWithEventingType = EventsContextApi<SubscribeEvents, PublishEvents>;
+  return <C extends ComponentWithEventingType>(Component: ComponentType<C>) => {
+    type ComponentWithOutEventingType = Subtract<C, ComponentWithEventingType>;
+
+    const ComponentWithEventing: FunctionComponent<ComponentWithOutEventingType> = (props: any) => {
+      useEffect(() => {
+        return function unsubscribe() {
+          if (subscription) {
+            subscription.unsubscribe();
+            subscription = null;
+          }
+          events.unsubscribe();
+        };
+      }, []);
+      let subscription: Unsubscribable | null = null;
+      const events = new Subject<AppEvent<SubscribeEvents>>();
+      const subscribe = (origin: string): Observable<AppEvent<SubscribeEvents>> => {
+        subscription = props.rootEvents
+          .pipe(
+            filter((event: AppEvent<any>) => {
+              if (event.origin === origin) {
+                return false;
+              }
+              const found = subscribeEvents.filter(value => value.name === event.name)[0];
```

Comment:
I would actually assume that changing what I am subscribed to could change during the component lifecycle (like be subscribed to some general event and start subscription to specific event after button click). With this I have to specify all the events in `withEvents` but will still need some filtering in the component later on to conditionally react to some events and not others.

Reply:
You're right, I went with the strong typed path. That is you have to declare all events that you want to `subscribe` to and all events you want to `publish`. We've never talked about a use case when we needed to change these dynamically but that could probably be solved with callbacks. Not a big fan of that though.